[WiP] Operator automatically handle tainted node

xumengpanda opened the pull request:
# Description

This is a WiP PR for discussion. 
Works to be done:
- [x] Add more unit tests in update_status_test and replace_failed_processes_test to improve coverage
- [ ] Consider to add a separate replacementTimeDuration to replace tainted node faster. Right now, cluster only replaces a failed process after 2 hours. The reaction time is too long for some taint key (e.g., maintenance or network disconnection).
- [ ] Consider to reduce the detection latency of a tainted node. Right now, the detection latency is the latency of operator's reconciliation invocation plus operator's reconciliation's execution time. Since reconciliation may not be invoked for a long time (hour), operator may not detect a 10min taint quickly enough. 

Future work
- [ ] Add end-to-end test in the kind-based end-to-end test framework
- [ ] Extend unit test framework with chaos scenarios and test the feature in failure scenarios in unit tests.

## Type of change

*Please select one of the options below.*

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Documentation
- Other

## Discussion

*Are there any design details that you would like to discuss further?*

## Testing

*Please describe the tests that you ran to verify your changes. Unit tests?
Manual testing?*

*Do we need to perform additional testing once this is merged, or perform in a larger testing environment?*

## Documentation

*Did you update relevant documentation within this repository?*

*If this change is adding new functionality, do we need to describe it in our user manual?*

*If this change is adding or removing subreconcilers, have we updated the core technical design doc to reflect that?*

*If this change is adding new safety checks or new potential failure modes, have we documented and how to debug potential issues?*

## Follow-up

*Are there any follow-up issues that we should pursue in the future?*

*Does this introduce new defaults that we should re-evaluate in the future?*

xumengpanda committed:
Add taint feature to auto replace tainted node

When a node is tainted by a key, operator will
mark pods on the node as NodeTaintDetected.
If a pods stay in  NodeTaintDetected for long enough, operator will
mark the pod as unhealthy and replace it with replacement logic.

SRE can configure the tainted keys operator should react to.
Each tainted key has a tainted duration which defines how long the pod
should stay in tainted mode before it is marked for replacement.

Change sets include
 * Set logger to output to stdout for replacements_test;
 * Add k8s node concept to unit test framework so that we can
   test node related features (e.g., taint a node, fail a node);
 * Update node and cluster config to propogate its effect to all reconcilation loop;
 * Update NodeTaintDetected condition time with node's tainted time; so that
   a pod is removed "closely" after configured duration after its tainted time;
 * Add taint tests in update_status_test.
   This test focuses on the testing the taint logic isolately;
 * Add taint test in replace_failed_process_groups_test.
   This test focuses on testing the interaction between taint function and other reconciliation loops

xumengpanda committed:
Add taint tests into update_status_test

xumengpanda committed:
Add test that replace a pod that is both NodeTaintDetected and NodeTaintReplacing with cluster reconciliation

xumengpanda committed:
Rafactor update_status_test

xumengpanda committed:
Add taint tests to replace_failed_processgroups

xumengpanda committed:
Test toggle taint feature

johscheuer commented on api/v1beta2/foundationdb_status.go:
```diff
@@ -134,6 +134,9 @@ type FoundationDBStatusProcessInfo struct {
 	// Excluded indicates whether the process has been excluded.
 	Excluded bool `json:"excluded,omitempty"`
 
+	// Tainted indicates whether the process's knode has been tainted for a while.
```
```suggestion
	// Tainted indicates whether the process's knode has been tainted for a while.
```

" for a while.", can we be more specific here? Are we talking about 1 second, 1 hour, 1 day?

Also I don't believe that the FoundationDB cluster status (https://apple.github.io/foundationdb/mr-status.html) has a field `tainted` in the process information. Why do we need this field here?

xumengpanda replied:
we don't need this.

johscheuer commented on api/v1beta2/foundationdbcluster_types.go:
```diff
@@ -609,7 +623,7 @@ func FilterByConditions(processGroupStatus []*ProcessGroupStatus, conditionRules
 }
 
-// MatchesConditions checks if the provided conditions are matching the current conditions of the process group.
+// MatchesConditions checks if the provided conditionRules exist in the process group's conditions
```
```suggestion
// MatchesConditions checks if the provided conditionRules match in the process group's conditions
```

exist is actually the wrong word here, since a conditionRule can also check that a condition is absent.

xumengpanda replied:
match creates the confusion that conditionRules are the same with ProcessGroupConditions. 
Is subset  a more precise word?

johscheuer replied:
I'm not sure if match creates confusion about the condition rules, as this represents best what those rules are meant for to match that either a condition exists or is absent.

xumengpanda replied:
I was confused when reading the comment.

johscheuer commented on api/v1beta2/foundationdbcluster_types.go:
```diff
@@ -701,6 +725,10 @@ const (
 	PodPending ProcessGroupConditionType = "PodPending"
 	// ReadyCondition is currently only used in the metrics.
 	ReadyCondition ProcessGroupConditionType = "Ready"
+	// NodeTaintDetected represents a pod's node is tainted but not long enough for operator to replace it
```
```suggestion
	// NodeTaintDetected represents a Pod's node is tainted but not long enough for operator to replace it
```

xumengpanda replied:
why pod has to be Pod?

johscheuer replied:
Because it's `Pod` and not `pod` -> https://kubernetes.io/docs/concepts/workloads/pods/

johscheuer commented on api/v1beta2/foundationdbcluster_types.go:
```diff
@@ -701,6 +725,10 @@ const (
 	ReadyCondition ProcessGroupConditionType = "Ready"
+	// NodeTaintDetected represents a pod's node is tainted but not long enough for operator to replace it
+	NodeTaintDetected ProcessGroupConditionType = "NodeTaintDetected"
+	// NodeTaintReplacing represents a pod whose node has been tainted for long enough and operator is replacing the pod
```
```suggestion
	// NodeTaintReplacing represents a Pod whose node has been tainted for long enough and operator is replacing the Pod
```

johscheuer commented on api/v1beta2/foundationdbcluster_types.go:
```diff
@@ -655,6 +669,16 @@ func (processGroupStatus *ProcessGroupStatus) GetConditionTime(conditionType Pro
 	return nil
 }
 
+func (processGroupStatus *ProcessGroupStatus) GetCondition(conditionType ProcessGroupConditionType) *ProcessGroupCondition {
```
Missing comment.

johscheuer commented on api/v1beta2/foundationdbcluster_types.go:
```diff
@@ -938,6 +972,20 @@ type MaintenanceModeOptions struct {
+//     durationInSeconds: 3600 # Ensure the taint is present for at least 1 hour before replacing Pods on a node with this taint
+type TaintReplacementOption struct {
+	// Tainted key
+	Key *string `json:"key,omitempty"`
```
Is there a reason this has to be a pointer?

xumengpanda replied:
I mimicked the code around it. 

IIUC, pointer allows the key be used immediately if it's changed (by human or other function). pls correct me if I'm wrong.

johscheuer replied:
I'm not sure I understand your comment. What do you mean by "used immediately"?

johscheuer commented on controllers/update_status.go:
```diff
@@ -624,6 +640,75 @@ func validateProcessGroup(ctx context.Context, r *FoundationDBClusterReconciler,
+						continue
+					}
+					if taint.Key == *taintConfiguredKey.Key || *taintConfiguredKey.Key == "*" {
+						hasValidTaint = true
+						logger.Info("MX Debug Info UpdateTaint", "Pod", pod.Name, "Add Condition", "NodeTaintDetected")
+						processGroupStatus.UpdateCondition(fdbv1beta2.NodeTaintDetected, true, cluster.Status.ProcessGroups, processGroupStatus.ProcessGroupID)
+						// Use node taint's timestamp as NodeTaintDetected condition's starting time
+						if taint.TimeAdded != nil && taint.TimeAdded.Time.Unix() < *processGroupStatus.GetConditionTime(fdbv1beta2.NodeTaintDetected) {
+							logger.Info("MX Debug Info Update NodeTaintDetected Time", "Pod", pod.Name,
+								"OriginalTime", processGroupStatus.GetConditionTime(fdbv1beta2.NodeTaintDetected),
+								"NewTime", taint.TimeAdded.Unix())
+							processGroupStatus.UpdateConditionTime(fdbv1beta2.NodeTaintDetected, taint.TimeAdded.Unix())
+						}
+						for _, processGroupCondition := range processGroupStatus.ProcessGroupConditions {
+							logger.Info("MX Debug Info Check to add NodeTaintReplacing Condition", "Pod", pod.Name,
+								"Now", time.Now().Unix(), "ConditionTimestamp", processGroupCondition.Timestamp,
+								"TaintConfiguredDuration", *taintConfiguredKey.DurationInSeconds)
+							if processGroupCondition.ProcessGroupConditionType == fdbv1beta2.NodeTaintDetected &&
+								time.Now().Unix()-processGroupCondition.Timestamp > *taintConfiguredKey.DurationInSeconds {
+								processGroupStatus.UpdateCondition(fdbv1beta2.NodeTaintReplacing, true, cluster.Status.ProcessGroups, processGroupStatus.ProcessGroupID)
+								logger.Info("Add NodeTaintReplacing condition", "Pod", pod.Name, "Node", node.Name,
+									"TaintKey", taint.Key, "TaintDetectedTime", processGroupCondition.Timestamp,
+									"TaintDuration", int64(time.Since(time.Unix(processGroupCondition.Timestamp, 0))),
+									"TaintValue", taint.Value, "TaintEffect", taint.Effect,
+									"NodeTaintDetectedConditionTimestamp", time.Unix(processGroupCondition.Timestamp, 0),
+									"Duration", time.Duration(*taintConfiguredKey.DurationInSeconds),
+									"Now", time.Now(),
+									"NodeTaintReplacingCondition", processGroupStatus.GetCondition(fdbv1beta2.NodeTaintDetected))
+							}
+						}
+					}
+				}
+			}
+			if !hasValidTaint {
+				// Remove NodeTaintDetected condition if the pod is no longer on a tainted node;
+				// This is needed especially when a node's status is flapping
+				// We do not remove NodeTaintDetected because the NodeTaintDetected condition is only added when taint is there for configured long time
+				processGroupStatus.UpdateCondition(fdbv1beta2.NodeTaintDetected, false, cluster.Status.ProcessGroups, processGroupStatus.ProcessGroupID)
+			}
+		}
```
It probably make sense to move this logic into a small method (this will make the testing easier and makes reading the method easier. The proposed design: https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/docs/design/node_taint_eviction.md#proposed-design proposed to query all nodes in a single call, if we are not doing this, we should update the design later to reflect the implementation, independent of the documentation for the new feature.

xumengpanda replied:
will update the design doc.

johscheuer commented on mock-kubernetes-client/client/client.go:
```diff
@@ -89,7 +89,12 @@ func NewMockClientWithHooks(scheme *runtime.Scheme, createHooks []func(ctx conte
 		}
 
-		return nil
+		node := corev1.Node{
```
Do you know what's the implication (test runtime) of adding the additional nodes is?

xumengpanda commented on controllers/replace_failed_process_groups_test.go:
```diff
@@ -63,6 +69,492 @@ var _ = Describe("replace_failed_process_groups", func() {
+						DurationInSeconds: &taintKeyMaintenanceDuration,
+					},
+				}
+				err = k8sClient.Update(context.TODO(), cluster)
```
need to get(cluster) object before making the change

xumengpanda committed:
Enable extra assertions and remove unnecessary comments

xumengpanda committed:
Add e2e test for taint feature

Mark it as pending test because our test infra does not allow test
to taint/access a k8s node.

xumengpanda commented:
Move to https://github.com/FoundationDB/fdb-kubernetes-operator/pull/1581
